ui-v2/src/routes/runs/index.tsx

```diff
@@ -1,9 +1,197 @@
+import { useSuspenseQueries } from "@tanstack/react-query";
 import { createFileRoute } from "@tanstack/react-router";
+import { zodValidator } from "@tanstack/zod-adapter";
+import { useCallback, useMemo } from "react";
+import { z } from "zod";
+import {
+	buildCountFlowRunsQuery,
+	buildPaginateFlowRunsQuery,
+	type FlowRunsPaginateFilter,
+} from "@/api/flow-runs";
+import { buildCountTaskRunsQuery } from "@/api/task-runs";
+import {
+	type PaginationState,
+	SORT_FILTERS,
+	type SortFilters,
+} from "@/components/flow-runs/flow-runs-list";
+import { RunsPage } from "@/components/runs/runs-page";
+
+const searchParams = z.object({
+	tab: z.enum(["flow-runs", "task-runs"]).optional().default("flow-runs"),
+	page: z.number().int().positive().optional().default(1).catch(1),
+	limit: z.number().int().positive().optional().default(10).catch(10),
+	sort: z.enum(SORT_FILTERS).optional().default("START_TIME_DESC"),
+	"hide-subflows": z.boolean().optional().default(false),
+});
+
+type SearchParams = z.infer<typeof searchParams>;
+
+const buildPaginationBody = (
+	search?: SearchParams,
+): FlowRunsPaginateFilter => ({
+	page: search?.page ?? 1,
+	limit: search?.limit ?? 10,
+	sort: search?.sort ?? "START_TIME_DESC",
+	flow_runs: search?.["hide-subflows"]
+		? {
+				operator: "and_",
+				parent_task_run_id: { operator: "and_", is_null_: true },
+			}
+		: undefined,
+});
 
 export const Route = createFileRoute("/runs/")({
+	validateSearch: zodValidator(searchParams),
 	component: RouteComponent,
+	loaderDeps: ({ search }) => buildPaginationBody(search),
+	loader: async ({ deps, context }) => {
+		// Await count queries - these don't change on pagination so won't cause suspense issues
+		const flowRunsCountResult = await context.queryClient.ensureQueryData(
+			buildCountFlowRunsQuery(),
+		);
+
+		const taskRunsCountResult = await context.queryClient.ensureQueryData(
+			buildCountTaskRunsQuery(),
+		);
```

Comment:
These don't need to be awaited. Awaiting them will slow down rendering.
```suggestion
		// Await count queries - these don't change on pagination so won't cause suspense issues
		void context.queryClient.prefetchQuery(
			buildCountFlowRunsQuery(),
		);

		void context.queryClient.prefectQuery(
			buildCountTaskRunsQuery(),
		);
```
The results also don't need to be returned from the loader.